File: backv2-github/DevDocs/backend/agents/HebrewOCRAgent.js
/**
 * Hebrew OCR Agent
 * 
 * Specialized agent for extracting text from documents with Hebrew content.
 * Uses Tesseract OCR with Hebrew language support and custom preprocessing
 * to improve recognition accuracy for Hebrew text.
 */

const { createWorker } = require('tesseract.js');
const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const sharp = require('sharp');
const logger = require('../utils/logger');
const config = require('../config');
const { v4: uuidv4 } = require('uuid');
const supabase = require('../db/supabase');
const storageService = require('../services/storage/supabaseStorageService');

// Promisify fs functions
const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);
const mkdir = promisify(fs.mkdir);
const unlink = promisify(fs.unlink);

/**
 * Hebrew OCR Agent class
 */
class HebrewOCRAgent {
  /**
   * Create a new HebrewOCRAgent
   * @param {Object} options - Agent options
   */
  constructor(options = {}) {
    this.options = {
      languages: ['heb', 'eng'],
      preprocess: true,
      oem: 3, // OCR Engine Mode: 0 = Legacy, 1 = Neural LSTM, 2 = Legacy + LSTM, 3 = Default
      psm: 6, // Page Segmentation Mode: 6 = Assume a single uniform block of text
      ...options
    };
    
    this.tempDir = path.join(config.upload.tempDir, 'ocr');
    
    // Create temp directory if it doesn't exist
    if (!fs.existsSync(this.tempDir)) {
      fs.mkdirSync(this.tempDir, { recursive: true });
    }
    
    logger.info('HebrewOCRAgent initialized');
  }
  
  /**
   * Process an image with OCR
   * @param {string|Buffer} input - Image file path or buffer
   * @param {Object} options - Processing options
   * @returns {Promise<Object>} OCR results
   */
  async processImage(input, options = {}) {
    const processingOptions = {
      ...this.options,
      ...options
    };
    
    try {
      // Get image buffer
      let imageBuffer;
      
      if (typeof input === 'string') {
        // Input is a file path
        imageBuffer = await readFile(input);
      } else if (Buffer.isBuffer(input)) {
        // Input is already a buffer
        imageBuffer = input;
      } else {
        throw new Error('Invalid input type. Expected file path or buffer.');
      }
      
      // Preprocess image if enabled
      if (processingOptions.preprocess) {
        imageBuffer = await this.preprocessImage(imageBuffer, processingOptions);
      }
      
      // Save preprocessed image for debugging if needed
      if (processingOptions.savePreprocessed) {
        const preprocessedPath = path.join(this.tempDir, `${uuidv4()}_preprocessed.png`);
        await writeFile(preprocessedPath, imageBuffer);
        logger.debug(`Saved preprocessed image to ${preprocessedPath}`);
      }
      
      // Process image with Tesseract
      const result = await this.recognizeText(imageBuffer, processingOptions);
      
      return result;
    } catch (error) {
      logger.error('Error processing image with OCR:', error);
      throw error;
    }
  }
  
  /**
   * Preprocess image to improve OCR accuracy
   * @param {Buffer} imageBuffer - Image buffer
   * @param {Object} options - Preprocessing options
   * @returns {Promise<Buffer>} Preprocessed image buffer
   */
  async preprocessImage(imageBuffer, options = {}) {
    try {
      // Default preprocessing options
      const preprocessOptions = {
        grayscale: true,
        binarize: true,
        threshold: 128,
        denoise: true,
        deskew: true,
        resize: false,
        width: 0,
        height: 0,
        ...options
      };
      
      // Start with sharp instance
      let image = sharp(imageBuffer);
      
      // Get image metadata
      const metadata = await image.metadata();
      
      // Convert to grayscale
      if (preprocessOptions.grayscale) {
        image = image.grayscale();
      }
      
      // Apply thresholding (binarization)
      if (preprocessOptions.binarize) {
        image = image.threshold(preprocessOptions.threshold);
      }
      
      // Apply denoising
      if (preprocessOptions.denoise) {
        image = image.median(1);
      }
      
      // Resize image if needed
      if (preprocessOptions.resize && preprocessOptions.width > 0 && preprocessOptions.height > 0) {
        image = image.resize(preprocessOptions.width, preprocessOptions.height, {
          fit: 'inside',
          withoutEnlargement: true
        });
      }
      
      // Apply deskewing if needed (not directly supported by sharp)
      // For deskewing, we would need a more complex algorithm
      
      // Get processed image buffer
      const processedBuffer = await image.toBuffer();
      
      return processedBuffer;
    } catch (error) {
      logger.error('Error preprocessing image:', error);
      throw error;
    }
  }
  
  /**
   * Recognize text in an image using Tesseract OCR
   * @param {Buffer} imageBuffer - Image buffer
   * @param {Object} options - Recognition options
   * @returns {Promise<Object>} Recognition results
   */
  async recognizeText(imageBuffer, options = {}) {
    try {
      // Create worker
      const worker = await createWorker();
      
      // Initialize worker with languages
      await worker.loadLanguage(options.languages.join('+'));
      await worker.initialize(options.languages.join('+'));
      
      // Set recognition parameters
      await worker.setParameters({
        tessedit_ocr_engine_mode: options.oem,
        tessedit_pageseg_mode: options.psm,
        preserve_interword_spaces: 1,
        textord_heavy_nr: 1,
        textord_force_make_prop_words: 0,
        tessedit_char_whitelist: options.whitelist || '',
        tessedit_char_blacklist: options.blacklist || ''
      });
      
      // Recognize text
      const { data } = await worker.recognize(imageBuffer);
      
      // Terminate worker
      await worker.terminate();
      
      // Process results
      const result = {
        text: data.text,
        confidence: data.confidence,
        words: data.words || [],
        lines: this.extractLines(data),
        blocks: this.extractBlocks(data),
        hocr: data.hocr,
        tsv: data.tsv
      };
      
      // Apply post-processing
      if (options.postprocess) {
        result.text = this.postprocessText(result.text, options);
      }
      
      return result;
    } catch (error) {
      logger.error('Error recognizing text:', error);
      throw error;
    }
  }
  
  /**
   * Extract lines from OCR result
   * @param {Object} data - OCR result data
   * @returns {Array} Extracted lines
   */
  extractLines(data) {
    if (!data.lines) {
      // If lines are not provided, extract from text
      return data.text.split('\\n').map(line => ({
        text: line,
        confidence: data.confidence
      }));
    }
    
    return data.lines;
  }
  
  /**
   * Extract blocks from OCR result
   * @param {Object} data - OCR result data
   * @returns {Array} Extracted blocks
   */
  extractBlocks(data) {
    if (!data.blocks) {
      // If blocks are not provided, use the whole text as a single block
      return [{
        text: data.text,
        confidence: data.confidence,
        bbox: data.bbox || { x0: 0, y0: 0, x1: 0, y1: 0 }
      }];
    }
    
    return data.blocks;
  }
  
  /**
   * Post-process OCR text
   * @param {string} text - OCR text
   * @param {Object} options - Post-processing options
   * @returns {string} Post-processed text
   */
  postprocessText(text, options = {}) {
    // Default post-processing options
    const postprocessOptions = {
      removeExtraSpaces: true,
      fixRtl: true,
      fixNumbers: true,
      ...options
    };
    
    let processedText = text;
    
    // Remove extra spaces
    if (postprocessOptions.removeExtraSpaces) {
      processedText = processedText.replace(/\s+/g, ' ').trim();
    }
    
    // Fix RTL text direction issues
    if (postprocessOptions.fixRtl) {
      // Add RTL mark at the beginning of each line
      processedText = processedText.split('\n').map(line => {
        // Check if line contains Hebrew characters
        if (/[\u0590-\u05FF]/.test(line)) {
          return '\u200F' + line;
        }
        return line;
      }).join('\n');
    }
    
    // Fix number recognition issues
    if (postprocessOptions.fixNumbers) {
      // Replace common OCR errors in numbers
      processedText = processedText
        .replace(/o/g, '0')
        .replace(/O/g, '0')
        .replace(/l/g, '1')
        .replace(/I/g, '1')
        .replace(/Z/g, '2')
        .replace(/S/g, '5')
        .replace(/B/g, '8');
    }
    
    return processedText;
  }
  
  /**
   * Process a document with OCR
   * @param {string} documentId - Document ID
   * @param {Object} options - Processing options
   * @returns {Promise<Object>} Processing results
   */
  async processDocument(documentId, options = {}) {
    try {
      // Get document from database
      const client = supabase.getClient();
      const { data: document, error } = await client
        .from('documents')
        .select('*')
        .eq('id', documentId)
        .single();
      
      if (error) {
        logger.error('Error getting document:', error);
        throw new Error('Error getting document');
      }
      
      if (!document) {
        throw new Error('Document not found');
      }
      
      // Check if document has storage metadata
      if (!document.metadata || !document.metadata.storage || !document.metadata.storage.path) {
        throw new Error('Document storage information not found');
      }
      
      // Get file from storage
      const fileBuffer = await storageService.downloadFile(document.metadata.storage.path);
      
      // Create a unique processing ID
      const processingId = uuidv4();
      
      // Create processing directory
      const processingDir = path.join(this.tempDir, processingId);
      await mkdir(processingDir, { recursive: true });
      
      // Save file to processing directory
      const filePath = path.join(processingDir, path.basename(document.metadata.storage.path));
      await writeFile(filePath, fileBuffer);
      
      // Process file based on type
      let result;
      
      if (document.file_type === '.pdf') {
        // Process PDF
        result = await this.processPdf(filePath, options);
      } else if (['.jpg', '.jpeg', '.png', '.tif', '.tiff', '.bmp'].includes(document.file_type)) {
        // Process image
        result = await this.processImage(filePath, options);
      } else {
        throw new Error(`Unsupported file type: ${document.file_type}`);
      }
      
      // Clean up processing directory
      try {
        await unlink(filePath);
        fs.rmdirSync(processingDir);
      } catch (cleanupError) {
        logger.warn(`Error cleaning up processing directory: ${cleanupError.message}`);
      }
      
      // Update document with OCR results
      await client
        .from('document_data')
        .insert({
          id: uuidv4(),
          document_id: documentId,
          data_type: 'ocr',
          content: result,
          created_at: new Date().toISOString(),
          updated_at: new Date().toISOString()
        });
      
      // Update document processing status
      await client
        .from('documents')
        .update({
          processing_status: 'ocr_completed',
          updated_at: new Date().toISOString()
        })
        .eq('id', documentId);
      
      return result;
    } catch (error) {
      logger.error('Error processing document with OCR:', error);
      
      // Update document with error
      try {
        const client = supabase.getClient();
        await client
          .from('documents')
          .update({
            processing_status: 'ocr_failed',
            processing_error: error.message,
            updated_at: new Date().toISOString()
          })
          .eq('id', documentId);
      } catch (updateError) {
        logger.error('Error updating document status:', updateError);
      }
      
      throw error;
    }
  }
  
  /**
   * Process a PDF document with OCR
   * @param {string} filePath - PDF file path
   * @param {Object} options - Processing options
   * @returns {Promise<Object>} Processing results
   */
  async processPdf(filePath, options = {}) {
    try {
      // TODO: Implement PDF processing with OCR
      // This would typically involve:
      // 1. Convert PDF pages to images
      // 2. Process each image with OCR
      // 3. Combine results
      
      // For now, return a mock result
      return {
        status: 'success',
        message: 'PDF processing not yet implemented',
        pages: []
      };
    } catch (error) {
      logger.error('Error processing PDF with OCR:', error);
      throw error;
    }
  }
}

module.exports = HebrewOCRAgent;

File: backv2-github/DevDocs/backend/agents/hebrew_ocr_agent.py
"""
Hebrew OCR Agent for accurate text recognition in Hebrew financial documents.
"""
import pytesseract
from PIL import Image
import numpy as np
import cv2
import os
import tempfile
from pathlib import Path
from typing import Dict, Any, List, Optional, Union
from .base_agent import BaseAgent
from ..utils.ocr_processor import OCRProcessor

class HebrewOCRAgent(BaseAgent):
    """Agent specialized in recognizing Hebrew text in financial documents."""

    def __init__(
        self,
        api_key: Optional[str] = None,
        **kwargs
    ):
        """
        Initialize the Hebrew OCR agent.

        Args:
            api_key: OpenRouter API key (not used for this agent)
            **kwargs: Additional parameters for the base agent
        """
        super().__init__(name="Hebrew OCR Agent")
        self.description = "I specialize in recognizing Hebrew text in financial documents with high accuracy."

        # OCR configuration
        self.lang = "heb+eng"  # Support for Hebrew and English
        self.config = "--psm 6 --oem 1"  # Tesseract configuration

        # Financial keywords in Hebrew for improving recognition accuracy
        self.financial_keywords = [
            "סה\"כ", "סך הכל", "חשבון", "תיק", "השקעות", "ריבית",
            "מניות", "אג\"ח", "תשואה", "דיבידנד", "עמלה", "מט\"ח",
            "יתרה", "הפקדה", "משיכה", "שער", "עו\"ש", "ני\"ע"
        ]

    def process(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """
        Process a task to extract text from an image or PDF.

        Args:
            task: Task dictionary with the following keys:
                - image_path: Path to the image file
                - or
                - image: Image as numpy array
                - or
                - pdf_path: Path to the PDF file
                - with_positions: Whether to include text positions (default: False)
                - language: OCR language (default: "heb+eng")
                - financial_doc: Whether this is a financial document (default: True)

        Returns:
            Dictionary with extracted text
        """
        # Get options
        with_positions = task.get('with_positions', False)
        language = task.get('language', self.lang)
        financial_doc = task.get('financial_doc', True)

        # Check if this is a PDF file
        if 'pdf_path' in task:
            pdf_path = task['pdf_path']
            if not isinstance(pdf_path, str):
                return {
                    'status': 'error',
                    'message': 'pdf_path must be a string'
                }

            # Get output path
            output_path = task.get('output_path', None)

            try:
                # Process the PDF with OCRmyPDF
                if financial_doc:
                    processed_pdf_path = OCRProcessor.process_financial_pdf(pdf_path, output_path)
                else:
                    processed_pdf_path = OCRProcessor.process_pdf(
                        pdf_path,
                        output_path,
                        language=language,
                        deskew=True,
                        clean=True
                    )

                # Extract text from the processed PDF
                text = OCRProcessor.extract_text_from_pdf(processed_pdf_path)

                return {
                    'status': 'success',
                    'text': text,
                    'processed_pdf_path': processed_pdf_path,
                    'ocr_options': {
                        'language': language,
                        'financial_doc': financial_doc
                    }
                }
            except Exception as e:
                return {
                    'status': 'error',
                    'message': f'Failed to process PDF: {str(e)}'
                }

        # Get the image
        if 'image_path' in task:
            image_path = task['image_path']
            if isinstance(image_path, str) or isinstance(image_path, Path):
                try:
                    img = Image.open(image_path)
                    img_array = np.array(img)
                except Exception as e:
                    return {
                        'status': 'error',
                        'message': f'Failed to read image from {image_path}: {str(e)}'
                    }
            else:
                return {
                    'status': 'error',
                    'message': 'image_path must be a string or Path object'
                }
        elif 'image' in task:
            img_array = task['image']
            if not isinstance(img_array, np.ndarray):
                return {
                    'status': 'error',
                    'message': 'image must be a numpy array'
                }
        else:
            return {
                'status': 'error',
                'message': 'Task must contain either image_path, image, or pdf_path'
            }

        # We already got the options above

        # Process the image
        if with_positions:
            result = self.extract_text_with_positions(img_array)
            return {
                'status': 'success',
                'text_with_positions': result
            }
        else:
            text = self.extract_text(img_array)
            return {
                'status': 'success',
                'text': text
            }

    def extract_text(self, image) -> str:
        """
        Extract text from an image with emphasis on Hebrew text.

        Args:
            image: Input image (path, PIL Image, or numpy array)

        Returns:
            Extracted text
        """
        # Load the image
        if isinstance(image, str) or isinstance(image, Path):
            img = Image.open(image)
            img_array = np.array(img)
        else:
            # If already received as numpy array
            img_array = image

        # Additional enhancement for financial documents
        enhanced = self._enhance_for_finance(img_array)

        # Run OCR
        text = pytesseract.image_to_string(enhanced, lang=self.lang, config=self.config)

        # Post-process the text
        processed_text = self._postprocess_text(text)

        return processed_text

    def extract_text_with_positions(self, image) -> List[Dict[str, Any]]:
        """
        Extract text with position in the image - important for document structure recognition.

        Args:
            image: Input image (path, PIL Image, or numpy array)

        Returns:
            List of dictionaries with text and position information
        """
        if isinstance(image, str) or isinstance(image, Path):
            img = Image.open(image)
            img_array = np.array(img)
        else:
            img_array = image

        enhanced = self._enhance_for_finance(img_array)

        # Get text data with position
        data = pytesseract.image_to_data(
            enhanced, lang=self.lang, config=self.config,
            output_type=pytesseract.Output.DICT
        )

        # Create list of words with positions
        text_with_positions = []
        for i in range(len(data['text'])):
            if int(data['conf'][i]) > 60:  # Only text above certain confidence level
                text_with_positions.append({
                    'text': data['text'][i],
                    'x': data['left'][i],
                    'y': data['top'][i],
                    'width': data['width'][i],
                    'height': data['height'][i],
                    'conf': data['conf'][i],
                    'block_num': data['block_num'][i],
                    'par_num': data['par_num'][i],
                    'line_num': data['line_num'][i]
                })

        return text_with_positions

    def _enhance_for_finance(self, img_array):
        """
        Enhance image specifically for financial text recognition.

        Args:
            img_array: Input image as numpy array

        Returns:
            Enhanced image
        """
        # Convert to grayscale if needed
        if len(img_array.shape) == 3:
            gray = cv2.cvtColor(img_array, cv2.COLOR_BGR2GRAY)
        else:
            gray = img_array.copy()

        # Increase contrast - effective for financial documents
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8,8))
        enhanced = clahe.apply(gray)

        # Use filter to improve small text (common in financial tables)
        enhanced = cv2.GaussianBlur(enhanced, (3, 3), 0)
        enhanced = cv2.addWeighted(gray, 1.5, enhanced, -0.5, 0)

        return enhanced

    def _postprocess_text(self, text):
        """
        Additional processing to improve recognized text.

        Args:
            text: OCR-recognized text

        Returns:
            Processed text
        """
        # Fix spacing
        processed = ' '.join(text.split())

        # Handle specific replacements for common OCR errors in Hebrew
        replacements = {
            'ו1': 'ני',
            '11"ס': 'ני"ע',
            'סהי"נ': 'סה"כ',
            'טר\'ח': 'מט"ח'
        }

        for old, new in replacements.items():
            processed = processed.replace(old, new)

        # Improve keyword recognition
        for keyword in self.financial_keywords:
            # Check if there's a similar but not identical word to a keyword
            for word in processed.split():
                if self._similar_words(word, keyword) and word != keyword:
                    processed = processed.replace(word, keyword)

        return processed

    def _similar_words(self, word1, word2, threshold=0.7):
        """
        Check if two words are similar (Levenshtein distance).

        Args:
            word1: First word
            word2: Second word
            threshold: Similarity threshold

        Returns:
            True if words are similar, False otherwise
        """
        # For financial documents, we can use a simpler similarity check
        # if there are many identical letters and similar length
        if abs(len(word1) - len(word2)) > 2:
            return False

        # Calculate number of common characters
        common_chars = set(word1) & set(word2)
        similarity = len(common_chars) / max(len(set(word1)), len(set(word2)))

        return similarity > threshold
